### ISSUE TITLE
Enable shuffle, mosaic, mixup if rect and all images are same size

### ISSUE THREAD

**Issue by deanmark** [ultralytics/ultralytics#8601]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Currently shuffle, mosaic and mixup are disabled for rect images during training. But if all the images are the same size, then these properties can be safely enabled, even though the images are rectangular. This will improve the performance on such datasets.

### Use case

This feature can be used for all rectangular training sets, where all images have the same size. This is relevant for all datasets that are composed of images from the same sensor, which outputs the same image sizes, e.g. the KITTI dataset where all images are 1382 x 512 pixels.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@deanmark hello! Thanks for bringing up this interesting point. 🌟

You're correct that if all images in a dataset are of the same size, enabling shuffle, mosaic, and mixup during training—even with rectangular images—shouldn't pose a problem and could indeed enhance model performance. This adjustment could be particularly beneficial for datasets like KITTI, where uniform image dimensions are a given.

For those willing to implement this feature, a conditional check could be added to verify the uniformity of image sizes before enabling these augmentations. Here's a simplified example of how this might look in code:

```python
if all_images_same_size(dataset):
    enable_augmentations(['shuffle', 'mosaic', 'mixup'])
```

This is a simplified representation, and the actual implementation would need to integrate seamlessly with the existing data loading and augmentation pipeline.

We're thrilled to hear you're interested in contributing by submitting a PR! Your initiative is greatly appreciated, and we look forward to seeing your proposed solution. Don't hesitate to reach out if you need any guidance or have further questions. Happy coding! 😊

**Comment by deanmark** [comment]
@glenn-jocher 
It's an easy fix to add this capability directly, just changing a few lines of code. But I'm thinking maybe this isn't the best way to go. KITTI for example has several sizes in the training set: [(1238, 374), (1224, 370), (1241, 376), (1242, 375)]. Thus the change as outlined above won't work with the KITTI dataset. It will only work if the user normalizes all the sizes in advance to the same rectangular shape, thus having limited applicability.

Probably the best way to handle this is by allowing the user to input an explicit (w,h) tuple for imgsz for training. In this case shuffle, mosaic and mixup should work without a problem. Currently it's only possible to set imgsz as the length of the largest dimension, with the program trying to keep the aspect ratio of the images while setting the size of the other dimension:

https://github.com/ultralytics/ultralytics/blob/36408c974c844f9885ff50320ebfa32162828d32/ultralytics/data/base.py#L161C1-L166C80

But changing the code to allow an explicit imgsz tuple is a much larger change in the codebase, and might have other implications I can't foresee.

What are your thoughts on the matter? Should I upload the small changes as outlined above? Or try to change the code to accept imgsz as a tuple?

**Comment by glenn-jocher** [comment]
Hey @deanmark! 👋

Thanks for diving deep into this and sharing your insights! 🚀 Your commitment to enhancing YOLOv8 is truly appreciated.

You've raised a valid point regarding the variability in image sizes within datasets like KITTI, and the idea of allowing an explicit `(w, h)` tuple for `imgsz` during training is intriguing. It indeed offers a more flexible approach to handling datasets with minor size disparities and could potentially streamline the augmentation process.

Given the complexity and the broader implications of such a change, it might be worth discussing further before diving into implementation. Your initial commit provides a good starting point for this conversation, and it's great to see proactive community contributions like yours!

For now, I'd suggest holding off on submitting the PR until we've had a chance to consider the best approach. It's important to weigh the benefits against the potential impact on the codebase and existing workflows.

Let's keep the dialogue open on this topic. Your thoughts and contributions are invaluable as we strive to make YOLOv8 even better. Thanks again for your proactive engagement! 🌟

**Comment by PelkiuBebras** [comment]
Hey,  @deanmark, @glenn-jocher.
This indeed seems like a great feature to enhance the accuracy of datasets, that have images of aspect ratios 1:2, 1:3, etc. I my case aspect ratio of images varies from 1:3 to 1:4. I would very appreciate having mosaic augmentation. As the aspect ratio and shape varies, explicitly defining input rectangular input height and width seems like a great option

**Comment by deanmark** [comment]
Hi @PelkiuBebras,
If you can create a version of your dataset where all images have same rectangular size, you can use my [commit](https://github.com/deanmark/ultralytics/commit/4848c978aca6a40adfecc28e2a25a09ed02186c8) to train with shuffle, mosaic and mixup enabled. 
Please check it out, and report back if this helps you.

**Comment by PelkiuBebras** [comment]
@deanmark,
Great idea. Will try it out

**Comment by nukithelegend** [comment]
Will this be implemented?

**Comment by glenn-jocher** [comment]
Thank you for your interest! The implementation is under discussion, and any updates will be shared in the repository. Please stay tuned for further developments.

**Comment by UltralyticsAssistant** [comment]
A potential fix is now merged in [PR #23419: “Disable shuffle only if batch shapes mismatch with `rect=True`”](https://github.com/ultralytics/ultralytics/pull/23419).

### What changed (and why it may resolve this)
The DataLoader shuffle logic for YOLO detection training was previously overly strict: when `rect=True` and `mode=="train"`, shuffling was always disabled and a warning was shown.  
With this update, shuffling is **only** disabled when it’s truly incompatible—specifically when `rect=True` **and** the dataset uses **multiple batch shapes** (i.e., `batch_shapes` vary across batches). This should:
- prevent shuffle from being turned off unnecessarily,
- reduce warning noise,
- improve training behavior/UX in `rect=True` setups where a single consistent batch shape is used.

### Credit
Merged thanks to @Y-T-G, with contributions from @glenn-jocher.

### How to test the fix
No specific version number was referenced in the PR, so the easiest ways to verify are:

- Install from the latest `main`:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`

- Or upgrade normally and wait for the next official PyPI release:
  - `pip install -U ultralytics`

After updating, re-run the same training command that triggered the issue and confirm whether:
- shuffling remains enabled when `rect=True` and batch shapes are consistent, and/or
- the “incompatible with shuffle” warning only appears when batch shapes actually vary.

### Feedback requested
Confirmation on whether this change resolves the issue in the original reproduction would be really helpful. If it still reproduces, sharing the training command, dataset setup details (especially around rectangular batching), and logs from the warning would help narrow down what’s left.

Thanks 🙏 for reporting this, and please share any further feedback if anything still seems off.

**Comment by deanmark** [comment]
Hi @glenn-jocher!
Congrats on fixing the shuffle! But the mosaic and mixup are still disabled on rect images, and can be safely enabled when all images are the same size. So the issue isn't completely closed.

**Comment by Y-T-G** [comment]
The reason it's disabled for mosaic is because mosaic produces square images by tiling 4 images into equal quadrants which defeats the point of `rect`.

**Comment by deanmark** [comment]
Mosaic can produce rectangular images, with 4 equal quadrants, by tiling 4 rectangular images. Just need to disable the checks that prevent it from running. It works. I created a branch once with the necessary changes - please have a look: https://github.com/deanmark/ultralytics/commit/4848c978aca6a40adfecc28e2a25a09ed02186c8

**Comment by glenn-jocher** [comment]
Thanks — your patch shows it can be forced on, but today our pipeline is built around a square mosaic canvas from a scalar `imgsz` (see the current [`Mosaic` implementation](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.Mosaic)), and we also explicitly zero `hyp.mosaic`/`hyp.mixup`/`hyp.cutmix` when `rect=True` (see [`YOLODataset.build_transforms()`](https://docs.ultralytics.com/reference/data/dataset/#ultralytics.data.dataset.YOLODataset.build_transforms)); making “rect + mosaic” a supported feature would likely need end-to-end `(h, w)` `imgsz` plumbing (augment + resize + batch-shape logic + tests).
